tests/test_strat.py

```diff
@@ -101,18 +105,113 @@ def check_efficiency():
+
+    def func(inputs, modules):
+        """Outputs gradient norm squared."""
+        y0, sde = inputs[0], modules[0]
+        ys = sdeint_adjoint(sde, y0, ts, bm, method="midpoint")
+        return (ys[-1] ** 2).sum(dim=1).mean(dim=0)
+
+    swiss_knife_gradcheck(func, y0, sde, eps=1e-7, rtol=1e-3, atol=1e-3, grad_params=True, gradgrad_params=True)
+
+
+def swiss_knife_gradcheck(func: Callable,
+                          inputs: TensorOrTensors,
+                          modules: Optional[ModuleOrModules] = (),
+                          eps: float = 1e-6,
+                          atol: float = 1e-5,
+                          rtol: float = 1e-3,
+                          grad_inputs=False,
+                          gradgrad_inputs=False,
+                          grad_params=False,
+                          gradgrad_params=False):
+    """Check grad and grad of grad wrt inputs and parameters of Modules.
+
+    Assumes func outputs a single scalar. Aimed to be as self-contained as
+    possible so that could be copied/pasted across projects.
+
+    Defaults for atol, rtol, and eps based on `torch.autograd.gradcheck`.
+    """
+
+    def convert_none_to_zeros(sequence, like_sequence):
+        return [torch.zeros_like(q) if p is None else p for p, q in zip(sequence, like_sequence)]
+
+    def flatten(sequence):
+        return torch.cat([p.reshape(-1) for p in sequence]) if len(sequence) > 0 else torch.tensor([])
+
+    if isinstance(inputs, torch.Tensor):
+        inputs = (inputs,)
+
+    if isinstance(modules, nn.Module):
+        modules = (modules,)
+
+    # Don't modify original objects.
+    modules = tuple(copy.deepcopy(m) for m in modules)
+    inputs = tuple(i.clone().requires_grad_() for i in inputs)
+    func_only_inputs = lambda *args: func(args, modules)  # noqa
+
+    # Grad wrt inputs.
+    if grad_inputs:
+        torch.autograd.gradcheck(func_only_inputs, inputs, eps=eps, atol=atol, rtol=rtol)
+
+    # Grad of grad wrt inputs.
+    if gradgrad_inputs:
+        torch.autograd.gradgradcheck(func_only_inputs, inputs, eps=eps, atol=atol, rtol=rtol)
+
+    # Grad wrt params.
+    if grad_params:
+        params = [p for m in modules for p in m.parameters() if p.requires_grad]
+        loss = func(inputs, modules)
+        framework_grad = flatten(convert_none_to_zeros(torch.autograd.grad(loss, params, create_graph=True), params))
+
+        numerical_grad = []
+        for param in params:
+            flat_param = param.reshape(-1)
+            for i in range(len(flat_param)):
+                flat_param[i] += eps  # In-place.
+                plus_eps = func(inputs, modules)
+                flat_param[i] -= eps
+
+                flat_param[i] -= eps
+                minus_eps = func(inputs, modules)
+                flat_param[i] += eps
+
+                numerical_grad.append((plus_eps - minus_eps).detach() / (2 * eps))
+                del plus_eps, minus_eps
+        numerical_grad = torch.stack(numerical_grad)
```

Comment:
Can this computation be wrapped in a `torch.no_grad()`?

Reply:
I don't see that strictly necessary. Though, I updated where detach is placed.